engine/common/soundlib/snd_ogg_vorbis.c

```diff
@@ -0,0 +1,250 @@
+GNU General Public License for more details.
+*/
+
+#include "soundlib.h"
+#include "crtlib.h"
+#include "xash3d_mathlib.h"
+#include "ogg_filestream.h"
+#include <vorbis/codec.h>
+#include <vorbis/vorbisfile.h>
+#include <string.h>
+#include <stdint.h>
+
+typedef struct vorbis_streaming_ctx_s
+{
+	file_t *file;
+	OggVorbis_File vf;
+} vorbis_streaming_ctx_t;
+
+static size_t FS_ReadOggVorbis( void *ptr, size_t blockSize, size_t nmemb, void *datasource )
+{
+	vorbis_streaming_ctx_t *ctx = (vorbis_streaming_ctx_t*)datasource;
+	return g_fsapi.Read( ctx->file, ptr, blockSize * nmemb );
+}
+
+static int FS_SeekOggVorbis( void *datasource, int64_t offset, int whence )
+{
+	vorbis_streaming_ctx_t *ctx = (vorbis_streaming_ctx_t*)datasource;
+	return g_fsapi.Seek( ctx->file, offset, whence );
+}
+
+static long FS_TellOggVorbis( void *datasource )
+{
+	vorbis_streaming_ctx_t *ctx = (vorbis_streaming_ctx_t*)datasource;
+	return g_fsapi.Tell( ctx->file );
+}
+
+static const ov_callbacks ov_callbacks_membuf = {
+	OggFilestream_Read,
+	OggFilestream_Seek,
+	NULL,
+	OggFilestream_Tell
+};
+
+static const ov_callbacks ov_callbacks_fs = {
+	FS_ReadOggVorbis,
+	FS_SeekOggVorbis,
+	NULL,
+	FS_TellOggVorbis
+};
+
+/*
+=================================================================
+
+	Ogg Vorbis decompression & streaming
+
+=================================================================
+*/
+static void Sound_ScanVorbisComments( OggVorbis_File *vf )
+{
+	const char *value;
+	vorbis_comment *vc = ov_comment( vf, -1 );
+	if( vc )
+	{	
+		if(( value = vorbis_comment_query( vc, "LOOPSTART", 0 ))) {
+			sound.loopstart = Q_atoi( value );
+			SetBits( sound.flags, SOUND_LOOPED );
+		}
+		else if(( value = vorbis_comment_query( vc, "LOOP_START", 0 ))) {
+			sound.loopstart = Q_atoi( value );
+			SetBits( sound.flags, SOUND_LOOPED );
+		}
+	}
+}
+
+qboolean Sound_LoadOggVorbis( const char *name, const byte *buffer, fs_offset_t filesize )
+{
+	long ret;
+	int section;
+	size_t written = 0;
+	ogg_filestream_t file;
+	OggVorbis_File vorbisFile;
+	
+	if( !buffer )
+		return false;
+
+	OggFilestream_Init( &file, name, buffer, filesize );
+	if( ov_open_callbacks( &file, &vorbisFile, NULL, 0, ov_callbacks_membuf ) < 0 ) {
+		Con_DPrintf( S_ERROR "%s: failed to load (%s): file reading error\n", __func__, file.name );
+		return false;
+	}
+
+	vorbis_info *info = ov_info( &vorbisFile, -1 );
+	if( info->channels < 1 || info->channels > 2 ) {
+		Con_DPrintf( S_ERROR "%s: failed to load (%s): unsuppored channels count\n", __func__, file.name );
+		return false;
+	}
+
+	sound.channels = info->channels;
+	sound.rate = info->rate;
+	sound.width = 2; // always 16-bit PCM
+	sound.type = WF_PCMDATA;
+	sound.samples = ov_pcm_total( &vorbisFile, -1 );
+	sound.size = sound.samples * sound.width * sound.channels;
+	sound.wav = (byte *)Mem_Calloc( host.soundpool, sound.size );
+
+	SetBits( sound.flags, SOUND_RESAMPLE );
+	Sound_ScanVorbisComments( &vorbisFile );
```

Comment:
Just wanted to tell that it's kinda awesome that you look for comment field here, as it's commonly used for an extensions like this in other idtech based engines.